accounts/abi/bind/backends/simulated.go

```diff
@@ -130,6 +134,65 @@ func (b *SimulatedBackend) rollback() {
 	b.pendingState, _ = state.New(b.pendingBlock.Root(), b.blockchain.StateCache(), nil)
 }
 
+// Reorg simulates a chain reorg with parameters depth and length.
+// f can be used to insert, reorder and exclude transactions.
+// f receives a slice of Transactions where each slice entry contains the
+// transactions that were contained in the orhpaned block.
+// The slice that is passed to f has length depth, and the one returned
+// must have at least length length.
+// Panics if length is not greater than depth.
+func (b *SimulatedBackend) Reorg(ctx context.Context, depth, length int64, f Reorderer) error {
+	if depth <= 0 {
+		panic("depth must be greater than 0")
+	}
+	// The chain inserter always chooses the most difficult chain, so a
+	// shorter chain would not trigger a reorg.
+	if length <= depth {
+		panic("reorg length must be greater than depth")
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	// Prevent lost TXs, call Commit() first.
+	if len(b.pendingBlock.Transactions()) != 0 {
+		return errors.New("cannot orphan dirty pending block")
+	}
+	// parent at pending - 1 - depth.
+	parentN := new(big.Int).Sub(b.pendingBlock.Number(), big.NewInt(depth+1))
+	parent, err := b.blockByNumberNoLock(ctx, parentN)
+	if err != nil {
+		return fmt.Errorf("retrieving reorg parent: %w", err)
+	}
+	// Collect orphaned TXs.
+	txs := make([]types.Transactions, depth)
+	for i := int64(0); i < depth; i++ {
+		blockN := new(big.Int).Add(parentN, big.NewInt(i+1))
+		block, _ := b.blockByNumberNoLock(ctx, blockN)
+		// Add the TXs from block parent + 1 + i.
+		txs[i] = block.Transactions()
+	}
+	// Modify the TXs with the Reorderer callback.
+	newTXs := f(txs)
+	if int64(len(newTXs)) > length {
+		panic("more new transactions than blocks")
+	}
+	// Generate new chain ontop of parent.
+	blocks, _ := core.GenerateChain(b.config, parent, ethash.NewFaker(), b.database, int(length), func(i int, block *core.BlockGen) {
+		// Add the modified TXs to their new blocks, if any.
+		if i < len(newTXs) {
+			for _, tx := range newTXs[i] {
+				block.AddTx(tx)
```

Comment:
Please use `block.AddTxWithChain(b.blockchain, tx)` otherwise the BLOCKHASH opcode will panic during execution.